File: 509. Fibonacci Number/509.FibonacciNumber-recursion.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    private Map<Integer, Integer> memo = new HashMap<>();

    public int fib(int targetN) {
        if (targetN == 0) return 0;
        if (targetN == 1) return 1;

        if (memo.containsKey(targetN)) return memo.get(targetN);

        int result = fib(targetN - 1) + fib(targetN - 2);
        memo.put(targetN, result);

        return result;
    }
}

File: 509. Fibonacci Number/509.FibonacciNumber-recursion.js
var fib = function (targetN, memo = {}) {
  if (targetN === 0) return 0;
  if (targetN === 1) return 1;

  if (memo[targetN]) return memo[targetN];

  memo[targetN] = fib(targetN - 1, memo) + fib(targetN - 2, memo);

  return memo[targetN];
};
